### PR TITLE
ar71xx: Add support for Teltonika RUT900

### PR DESCRIPTION
**Pull request by lairsdragon**
Teltonika RUT900 is a Router with LTE dual SIM, WiFi, 4x Ethernet
ports, I/O, RS232, RS485, GPS.

The device ist based on a Atheros AR9344 rev 3,

Specifications:
- 560/450/225 MHz (CPU/DDR/AHB)
- 128 MB of RAM
- 16 MB of FLASH
- Serial Console header on a Card Board edge connector
- 4x 10/100 Mbps Ethernet (3x LAN, 1x WAN)
- 2.4 GHz Wifi
- 2x external, detachable Wifi antennas
- LTE Modem Huawei ME909u-521 (Also other Modem seen)
- 2x LTE antennas
- 1x GPS antenna
- 7x LED, 1x button
- 1x USB Connector
- 1x Serial RS232
- 1x Serial RS485
- 1x MicroSD Card

The GPL sources of the device are available at www.teltonika.lt/gpl/
and are based on OpenWRT Barrier Breaker (14.07)

Running from tftp:
The Router starts into the uboot Webupdater if the Button ist pressed
more than 3 seconds, if no Network cable is attached it starts the
uboot serial console, from there the router loads the firmware image via
tftpboot from 192.168.1.2:firmware.bin (the router has the
192.168.1.1). With bootm the loaded image will be booted.

Signed-off-by: Steffen Weinreich <steve@weinreich.org>

### CODE REVIEW

**Comment by pepe2k** [comment]
Hello @lairsdragon,

Most important things after just a brief review:
1. I would prefer to use or modify an existing mtd partition parses (of course, if it's possible) instead of adding new one. Based on Teltonika GPL sources I see that they are using modified `tp-link` parser which is/was wrong (it breaks everything else). I will take a look and check if there is a way to modify `tp-link` parser in the way it can support also Teltonika partition layout.
2. Please, avoid commented out code.

I have RUT900 somewhere, give me few days, I will take a look at it more closely.

Cheers,
Piotr

**Comment by lairsdragon** [comment]
Hi @pepe2k 

Here some details from a RUT900:
*dmesg*: 
```[    0.000000] Linux version 3.10.36 (Mindaugas@Mindaugo) (gcc version 4.8.3 (OpenWrt/Linaro GCC 4.8-2014.01 r40569) ) #29 Wed Jun 29 10:27:51 EEST 2016
[    0.000000] MyLoader: sysp=511b6c54, boardp=e6f2066e, parts=9bcd964c
[    0.000000] bootconsole [early0] enabled
[    0.000000] CPU revision is: 0001974c (MIPS 74Kc)
[    0.000000] SoC: Atheros AR9344 rev 3
[    0.000000] Clocks: CPU:560.000MHz, DDR:450.000MHz, AHB:225.000MHz, Ref:40.000MHz
[    0.000000] Determined physical RAM map:
[    0.000000]  memory: 08000000 @ 00000000 (usable)
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x00000000-0x07ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x00000000-0x07ffffff]
[    0.000000] On node 0 totalpages: 32768
[    0.000000] free_area_init_node: node 0, pgdat 80300b60, node_mem_map 81000000
[    0.000000]   Normal zone: 256 pages used for memmap
[    0.000000]   Normal zone: 0 pages reserved
[    0.000000]   Normal zone: 32768 pages, LIFO batch:7
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, VIPT, cache aliases, linesize 32 bytes
[    0.000000] pcpu-alloc: s0 r0 d32768 u32768 alloc=1*32768
[    0.000000] pcpu-alloc: [0] 0 
[    0.000000] Built 1 zonelists in Zone order, mobility grouping on.  Total pages: 32512
[    0.000000] Kernel command line:  board=TLT-RUT900 console=ttyS0,115200 rootfstype=squashfs,jffs2 noinitrd lpj=1394688 quiet loglevel=3
[    0.000000] PID hash table entries: 512 (order: -1, 2048 bytes)
[    0.000000] Dentry cache hash table entries: 16384 (order: 4, 65536 bytes)
[    0.000000] Inode-cache hash table entries: 8192 (order: 3, 32768 bytes)
[    0.000000] Writing ErrCtl register=00000000
[    0.000000] Readback ErrCtl register=00000000
[    0.000000] Memory: 126316k/131072k available (2207k kernel code, 4756k reserved, 558k data, 240k init, 0k highmem)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS:51
[    0.000000] Calibrating delay loop (skipped) preset value.. 278.93 BogoMIPS (lpj=1394688)
[    0.000000] pid_max: default: 32768 minimum: 301
[    0.000000] Mount-cache hash table entries: 512
[    0.000000] NET: Registered protocol family 16
[    0.000000] MIPS: machine is TELTONIKA TLT-RUT900
[    0.410000] bio: create slab <bio-0> at 0
[    0.420000] pca953x 0-0074: interrupt support not compiled in
[    0.420000] i2c-gpio i2c-gpio.0: using pins 17 (SDA) and 16 (SCL)
[    0.420000] Switching to clocksource MIPS
[    0.420000] NET: Registered protocol family 2
[    0.420000] TCP established hash table entries: 1024 (order: 1, 8192 bytes)
[    0.420000] TCP bind hash table entries: 1024 (order: 0, 4096 bytes)
[    0.420000] TCP: Hash tables configured (established 1024 bind 1024)
[    0.420000] TCP: reno registered
[    0.420000] UDP hash table entries: 256 (order: 0, 4096 bytes)
[    0.420000] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes)
[    0.420000] NET: Registered protocol family 1
[    0.430000] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.430000] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.430000] msgmni has been set to 246
[    0.430000] io scheduler noop registered
[    0.430000] io scheduler deadline registered (default)
[    0.440000] Serial: 8250/16550 driver, 1 ports, IRQ sharing disabled
[    0.460000] serial8250.0: ttyS0 at MMIO 0x18020000 (irq = 11) is a 16550A
[    0.460000] console [ttyS0] enabled, bootconsole disabled
[    0.460000] ar934x-hs-uart.0: ttyATH0 at MMIO 0x18500000 (irq = 14) is a AR934X UART
[    0.460000] ath79-spi ath79-spi: master is unqueued, this is deprecated
[    0.460000] m25p80 spi0.0: found w25q128, expected m25p80
[    0.460000] m25p80 spi0.0: w25q128 (16384 Kbytes)
[    0.460000] 7 tp-link partitions found on MTD device spi0.0
[    0.460000] Creating 7 MTD partitions on "spi0.0":
[    0.460000] 0x000000000000-0x000000020000 : "u-boot"
[    0.460000] 0x000000020000-0x000000030000 : "config"
[    0.460000] 0x000000030000-0x000000040000 : "art"
[    0.470000] 0x000000040000-0x000000139a1c : "kernel"
[    0.470000] mtd: partition "kernel" must either start or end on erase block boundary or be smaller than an erase block -- forcing read-only
[    0.470000] 0x000000139a1c-0x000000f70000 : "rootfs"
[    0.470000] mtd: partition "rootfs" must either start or end on erase block boundary or be smaller than an erase block -- forcing read-only
[    0.470000] mtd: device 4 (rootfs) set to be root filesystem
[    0.470000] 1 squashfs-split partitions found on MTD device rootfs
[    0.470000] 0x0000008c0000-0x000000f70000 : "rootfs_data"
[    0.470000] 0x000000040000-0x000000f70000 : "firmware"
[    0.470000] 0x000000f70000-0x000001000000 : "event-log"
[    0.480000] libphy: ag71xx_mdio: probed
[    1.040000] ag71xx-mdio.1: Found an AR934X built-in switch
[    2.070000] eth0: Atheros AG71xx at 0xba000000, irq 5, mode:GMII
[    2.620000] ag71xx ag71xx.0: connected to PHY at ag71xx-mdio.1:04 [uid=004dd042, driver=Generic PHY]
[    2.620000] eth1: Atheros AG71xx at 0xb9000000, irq 4, mode:MII
[    2.630000] TCP: cubic registered
[    2.630000] NET: Registered protocol family 17
[    2.630000] Bridge firewalling registered
[    2.630000] 8021q: 802.1Q VLAN Support v1.8
[    2.640000] VFS: Mounted root (squashfs filesystem) readonly on device 31:4.
[    2.640000] Freeing unused kernel memory: 240K (80314000 - 80350000)
[    5.420000] usbcore: registered new interface driver usbfs
[    5.420000] usbcore: registered new interface driver hub
[    5.420000] usbcore: registered new device driver usb
[    5.430000] SCSI subsystem initialized
[    5.440000] ehci_hcd: USB 2.0 'Enhanced' Host Controller (EHCI) Driver
[    5.450000] usbcore: registered new interface driver usb-storage
[    6.960000] jffs2: notice: (273) jffs2_build_xattr_subsystem: complete building xattr subsystem, 1 of xdatum (0 unchecked, 0 orphan) and 15 of xref (0 dead, 0 orphan) found.
[    7.090000] jffs2: notice: (270) jffs2_build_xattr_subsystem: complete building xattr subsystem, 1 of xdatum (0 unchecked, 0 orphan) and 15 of xref (0 dead, 0 orphan) found.
[   10.260000] NET: Registered protocol family 10
[   10.300000] NTFS driver 2.1.30 [Flags: R/O MODULE].
[   10.320000] Initializing XFRM netlink socket
[   10.340000] NET: Registered protocol family 15
[   10.340000] tun: Universal TUN/TAP device driver, 1.6
[   10.340000] tun: (C) 1999-2004 Max Krasnyansky <maxk@qualcomm.com>
[   10.410000] l2tp_core: L2TP core driver, V2.0
[   10.410000] l2tp_netlink: L2TP netlink interface
[   10.420000] gre: GRE over IPv4 demultiplexor driver
[   10.420000] ip_gre: GRE over IPv4 tunneling driver
[   10.450000] nf_conntrack version 0.5.0 (1977 buckets, 7908 max)
[   10.460000] ip6_tables: (C) 2000-2006 Netfilter Core Team
[   10.720000] u32 classifier
[   10.720000]     input device check on
[   10.720000]     Actions configured
[   10.720000] Mirror/redirect action on
[   10.740000] usbcore: registered new interface driver cdc_acm
[   10.740000] cdc_acm: USB Abstract Control Model driver for USB modems and ISDN adapters
[   10.740000] usbcore: registered new interface driver cdc_wdm
[   10.760000] Loading modules backported from Linux version master-2014-11-04-0-gf3660a2
[   10.760000] Backport generated by backports.git backports-20141023-2-g4ff890b
[   10.800000] Ebtables v2.0 registered
[   10.820000] ip_tables: (C) 2000-2006 Netfilter Core Team
[   10.850000] Netfilter messages via NETLINK v0.30.
[   10.880000] usbcore: registered new interface driver usbserial
[   10.880000] usbcore: registered new interface driver usbserial_generic
[   10.880000] usbserial: USB Serial support registered for generic
[   10.970000] xt_time: kernel timezone is -0000
[   10.970000] usbcore: registered new interface driver cdc_ether
[   11.050000] usbcore: registered new interface driver ftdi_sio
[   11.050000] usbserial: USB Serial support registered for FTDI USB Serial Device
[   11.060000] cfg80211: Calling CRDA to update world regulatory domain
[   11.060000] cfg80211: World regulatory domain updated:
[   11.060000] cfg80211:  DFS Master region: unset
[   11.060000] cfg80211:   (start_freq - end_freq @ bandwidth), (max_antenna_gain, max_eirp), (dfs_cac_time)
[   11.060000] cfg80211:   (2402000 KHz - 2472000 KHz @ 40000 KHz), (N/A, 2000 mBm), (N/A)
[   11.060000] cfg80211:   (2457000 KHz - 2482000 KHz @ 40000 KHz), (N/A, 2000 mBm), (N/A)
[   11.060000] cfg80211:   (2474000 KHz - 2494000 KHz @ 20000 KHz), (N/A, 2000 mBm), (N/A)
[   11.060000] cfg80211:   (5170000 KHz - 5250000 KHz @ 80000 KHz), (N/A, 2000 mBm), (N/A)
[   11.060000] cfg80211:   (5250000 KHz - 5330000 KHz @ 80000 KHz, 160000 KHz AUTO), (N/A, 2000 mBm), (0 s)
[   11.060000] cfg80211:   (5490000 KHz - 5730000 KHz @ 160000 KHz), (N/A, 2000 mBm), (0 s)
[   11.060000] cfg80211:   (5735000 KHz - 5835000 KHz @ 80000 KHz), (N/A, 2000 mBm), (N/A)
[   11.060000] cfg80211:   (57240000 KHz - 63720000 KHz @ 2160000 KHz), (N/A, 0 mBm), (N/A)
[   11.130000] ctnetlink v0.93: registering with nfnetlink.
[   11.160000] PPP generic driver version 2.4.2
[   11.160000] PPP MPPE Compression module registered
[   11.160000] NET: Registered protocol family 24
[   11.190000] PPTP driver version 0.8.5
[   11.190000] usbcore: registered new interface driver qmi_wwan
[   11.260000] l2tp_ppp: PPPoL2TP kernel driver, V2.0
[   11.270000] usbcore: registered new interface driver option
[   11.270000] usbserial: USB Serial support registered for GSM modem (1-port)
[   11.300000] ath: EEPROM regdomain: 0x0
[   11.300000] ath: EEPROM indicates default country code should be used
[   11.300000] ath: doing EEPROM country->regdmn map search
[   11.300000] ath: country maps to regdmn code: 0x3a
[   11.300000] ath: Country alpha2 being used: US
[   11.300000] ath: Regpair used: 0x3a
[   11.340000] ieee80211 phy0: Selected rate control algorithm 'minstrel_ht'
[   11.360000] cfg80211: Calling CRDA for country: US
[   11.360000] cfg80211: Regulatory domain changed to country: US
[   11.360000] cfg80211:  DFS Master region: FCC
[   11.360000] cfg80211:   (start_freq - end_freq @ bandwidth), (max_antenna_gain, max_eirp), (dfs_cac_time)
[   11.360000] cfg80211:   (2402000 KHz - 2472000 KHz @ 40000 KHz), (N/A, 3000 mBm), (N/A)
[   11.360000] cfg80211:   (5170000 KHz - 5250000 KHz @ 80000 KHz, 160000 KHz AUTO), (N/A, 1700 mBm), (N/A)
[   11.360000] cfg80211:   (5250000 KHz - 5330000 KHz @ 80000 KHz, 160000 KHz AUTO), (N/A, 2300 mBm), (0 s)
[   11.360000] cfg80211:   (5735000 KHz - 5835000 KHz @ 80000 KHz), (N/A, 3000 mBm), (N/A)
[   11.360000] cfg80211:   (57240000 KHz - 63720000 KHz @ 2160000 KHz), (N/A, 4000 mBm), (N/A)
[   11.360000] ieee80211 phy0: Atheros AR9340 Rev:0 mem=0xb8100000, irq=47
[   11.430000] mmc_spi spi0.1: ASSUMING 3.2-3.4 V slot power
[   14.490000] mmc_spi spi0.1: SD/MMC host mmc0, no DMA, no WP, no poweroff
[   14.500000] mmc0: host doesn't support card's voltages
[   14.500000] mmc0: error -22 whilst initialising SDIO card
[   14.500000] mmc0: host doesn't support card's voltages
[   14.500000] mmc0: error -22 whilst initialising SD card
[   14.500000] mmc0: host doesn't support card's voltages
[   14.500000] mmc0: error -22 whilst initialising MMC card
[   18.980000] ehci-platform: EHCI generic platform driver
[   18.980000] ehci-platform ehci-platform: EHCI Host Controller
[   18.990000] ehci-platform ehci-platform: new USB bus registered, assigned bus number 1
[   19.000000] ehci-platform ehci-platform: TX-TX IDP fix enabled
[   19.010000] ehci-platform ehci-platform: irq 3, io mem 0x1b000000
[   19.030000] ehci-platform ehci-platform: USB 2.0 started, EHCI 1.00
[   19.040000] hub 1-0:1.0: USB hub found
[   19.040000] hub 1-0:1.0: 1 port detected
[   19.310000] jffs2: notice: (1107) jffs2_build_xattr_subsystem: complete building xattr subsystem, 0 of xdatum (0 unchecked, 0 orphan) and 0 of xref (0 dead, 0 orphan) found.
[   19.360000] usb 1-1: new high-speed USB device number 2 using ehci-platform
[   19.770000] hub 1-1:1.0: USB hub found
[   19.770000] hub 1-1:1.0: 3 ports detected
[   20.160000] usb 1-1.1: new full-speed USB device number 3 using ehci-platform
[   20.340000] ftdi_sio 1-1.1:1.0: FTDI USB Serial Device converter detected
[   20.340000] usb 1-1.1: Detected FT232RL
[   20.350000] usb 1-1.1: Number of endpoints 2
[   20.350000] usb 1-1.1: Endpoint 1 MaxPacketSize 16384
[   20.360000] usb 1-1.1: Endpoint 2 MaxPacketSize 16384
[   20.360000] usb 1-1.1: Setting MaxPacketSize 64
[   20.430000] usb 1-1.1: FTDI USB Serial Device converter now attached to ttyUSB0
[   21.330000] usb 1-1.3: new high-speed USB device number 4 using ehci-platform
[   21.480000] cdc_ether 1-1.3:2.0 eth2: register 'cdc_ether' at usb-ehci-platform-1.3, CDC Ethernet Device, 02:1e:10:1f:00:00
[   21.510000] option 1-1.3:2.2: GSM modem (1-port) converter detected
[   21.550000] usb 1-1.3: GSM modem (1-port) converter now attached to ttyUSB1
[   21.550000] option 1-1.3:2.3: GSM modem (1-port) converter detected
[   21.680000] usb 1-1.3: GSM modem (1-port) converter now attached to ttyUSB2
[   21.720000] option 1-1.3:2.4: GSM modem (1-port) converter detected
[   21.720000] usb 1-1.3: GSM modem (1-port) converter now attached to ttyUSB3
[   21.770000] option 1-1.3:2.5: GSM modem (1-port) converter detected
[   21.780000] usb 1-1.3: GSM modem (1-port) converter now attached to ttyUSB4
[   26.880000] device eth0 entered promiscuous mode
[   27.050000] cdc_ether 1-1.3:2.0 eth2: kevent 11 may have been dropped
[   29.240000] cfg80211: Calling CRDA for country: DE
[   29.250000] cfg80211: Regulatory domain changed to country: DE
[   29.250000] cfg80211:  DFS Master region: ETSI
[   29.260000] cfg80211:   (start_freq - end_freq @ bandwidth), (max_antenna_gain, max_eirp), (dfs_cac_time)
[   29.270000] cfg80211:   (2400000 KHz - 2483000 KHz @ 40000 KHz), (N/A, 2000 mBm), (N/A)
[   29.270000] cfg80211:   (5150000 KHz - 5250000 KHz @ 80000 KHz, 200000 KHz AUTO), (N/A, 2000 mBm), (N/A)
[   29.280000] cfg80211:   (5250000 KHz - 5350000 KHz @ 80000 KHz, 200000 KHz AUTO), (N/A, 2000 mBm), (0 s)
[   29.290000] cfg80211:   (5470000 KHz - 5725000 KHz @ 160000 KHz), (N/A, 2700 mBm), (0 s)
[   29.300000] cfg80211:   (57000000 KHz - 66000000 KHz @ 2160000 KHz), (N/A, 4000 mBm), (N/A)
```
*/proc/mtd*: 
```dev:    size   erasesize  name
mtd0: 00020000 00010000 "u-boot"
mtd1: 00010000 00010000 "config"
mtd2: 00010000 00010000 "art"
mtd3: 000f9a1c 00010000 "kernel"
mtd4: 00e365e4 00010000 "rootfs"
mtd5: 006b0000 00010000 "rootfs_data"
mtd6: 00f30000 00010000 "firmware"
mtd7: 00090000 00010000 "event-log"
```

**Comment by mkresin** [comment]
@pepe2k short note, the led naming doesn't follow the board:colour:function pattern.

**Comment by lairsdragon** [comment]
Hi @pepe2k, @mkresin 

- fix the led naming
- fix the commented out code in `mach-tlt-rut900.c`

fixing the partition magic is above my current skill set, i have to leave this for a more skilled person :-)

File: target/linux/ar71xx/base-files/etc/diag.sh [hunk 1]
```diff
@@ -422,6 +422,9 @@ get_status_led() {
 	tl-wdr6500-v2)
 		status_led="tp-link:white:system"
 		;;
+	tlt-rut900)
+		status_led="status_green"
```
Comment (mkresin) [inline target/linux/ar71xx/base-files/etc/diag.sh]: this need to be updated as well

Comment (lairsdragon) [inline target/linux/ar71xx/base-files/etc/diag.sh]: fixed in updated pull request

**Comment by lairsdragon** [comment]
Hi @pepe2k and @mkresin 

I've messed the things up by squashing the branch, now I have fixed it, add missing pieces like Kconfig.openwrt, machtypes.h etc.

sorry for the mess,

cheerio
    Steve

**Comment by blogic** [comment]
merge, Thanks ! (sorry for the delay)

**Comment by pepe2k** [comment]
@blogic

> merge, Thanks ! (sorry for the delay)

In my opinion, this PR wasn't ready to merge, it still has some issues pointed out above by me and @mkresin.

**Comment by Redfoxmoon3** [comment]
@lairsdragon any updates on this? Will you rebase it against the current master?

**Comment by lairsdragon** [comment]
Hi @Redfoxmoon3 

I have rebased the pull request to the current master, LED naming `diag.sh`  and commented out code in `mach-tlt-rut900.c` ist fixed.

Still open is the mtd partitioning, any hints how to integrate  this  into a single partitioning functing is very welcome.

cheerio Steve

**Comment by lairsdragon** [comment]
Hi @mkresin & @pepe2k 

What can I do further to get this pull request into master? 

cheerio

**Comment by lairsdragon** [comment]
Hi @mkresin & @pepe2k

Any update on this pull request?

cheerio
    Steve

**Comment by pepe2k** [comment]
Hello @lairsdragon,

My first comment is still valid - please, use static mtd partitions definition instead of that custom parser. If you don't know ho to do it, I can help but not faster than in next month.

Cheers,
Piotr

**Comment by lairsdragon** [comment]
Hi @pepe2k

The tltpart.c is mostly a copy of tplinkpart.c, even the author wasn't changed in the header.

The actual flash layout of tplink and tlt900 differs in multiple ways; tplink come with 5 partitions (uboot, kernel, rootfs, art, firmware) and tlt900 come with 7 partition (uboot, config, art, kernel, rootfs, firmware, event-log), they have even a different order on the flash:

```
rut900:                       tplink:            
                                                 
+-----------------+           +-----------------+
| uboot (128k)    |           | uboot (64/128k) |
+-----------------+           +------------+----+
| config (64k)    |           | kernel (?k)| Fi |
+-----------------+           :            : rm :
| art    (64k)    |           +------------+ wa +
+------------+----+           | rootfs (?k)| re |
| kernel (?k)| Fi |           :            :    :
:            : rm :           +-----------------+
+------------+ wa +           | art       (64k) |
| rootfs (?k)| re |           +-----------------+
:            :    :
+-----------------+
| event-log (576k)|
+-----------------+
```

One way to  reduce the code duplication between tltpart.c and tplinkpart.c could be to add only the tlt parser to tplinkpart.c:

``` c
  static struct mtd_part_parser tlt_parser = {
          .owner		= THIS_MODULE,
          .parse_fn	        = tlt_parse_partitions,
          .name		= "tlt-rut900",
  };

  static int __init tplink_parser_init(void)
  {
          register_mtd_parser(&tplink_parser);
          register_mtd_parser(&tplink_64k_parser);
          register_mtd_parser(&tlt_parser);	

          return 0;
  }
```
is this a better solution?

cheerio
     Steve

**Comment by pepe2k** [comment]
Hello @lairsdragon,

> is this a better solution?

I really don't think so. There is a parser for TP-Link because this vendor has devices with same layout on different FLASH sizes (4-16 MB), sometimes in same/similar model with just a different hardware version. Teltonika uses only one FLASH size in RUT9xx series (or I'm not aware of different versions?).
You really don't need a custom parser if you know all the partition offsets and sizes (for kernel+rootfs we use firmware partition, you wrote about 64/128 KB U-Boot partition size but in the code you have `define TLT_UBOOT_LEN 0x20000`, isn't this static? And what about this one: '`kernel_offset = TLT_UBOOT_LEN + TLT_CONFIG_LEN + TLT_ART_LEN`, doesn't this look like a static value for you?).

Cheers,
Piotr

**Comment by lairsdragon** [comment]
Hi @pepe2k 

> You really don't need a custom parser if you know all the partition offsets and sizes

Ahhhh, now I understood your point, define the static partition as `struct mtd_partition`  in `mach-tlt-rut900.c` and remove the custom parser. I was not aware of this variant, thanks.

I will rework the pull request accordingly

cheerio
    Steve

**Comment by pepe2k** [comment]
Hello @lairsdragon,

I suppose you could just use the `MTDPARTS` in https://github.com/lede-project/source/blob/master/target/linux/ar71xx/image/generic.mk

Cheers!
Piotr

**Comment by lairsdragon** [comment]
Hi @pepe2k 

Interesting, now I have learned three ways to define the flash layout in LEDE :-)

In the change I made yesterday, I have choosen  the `mtd_partion`  definition in the machine file `mach-tlt-rut900.c`. Does this fit into the LEDE source?

cheerio
    Steve

**Comment by lairsdragon** [comment]
Hi @pepe2k 

Is there anything else I should change in this PR?

cheerio
    Steve

**Comment by lairsdragon** [comment]
Hi @pepe2k , @mkresin 

Any chances to get this merged? Anything left I should change?

cheerio
    Steve

**Comment by blogic** [comment]
With the remerge in progress, all PRs on the lede-project organisation will be closed. Please help getting this merged or rebase/post it on the openwrt project page (https://github.com/openwrt/openwrt/pulls). All remaining PRs will be closed in 30 days.

**Comment by pepe2k** [comment]
Hi @lairsdragon,

I have just pushed support for RUT900 (https://github.com/lede-project/source/commit/1f06144cdec1487122201faa9350825ed565d99c). If you are still interested in adding support for your model (I suppose it's RUT955 or RUT905, definitely not a RUT900, based on your description and vendor datasheets), you can base on my work and just extend the mach-rut9xx.c file.

Cheers,
Piotr

**Pull request closed without merging**